### PR TITLE
Fixed issue #17992: unable to reimport tokens from old table

### PR DESCRIPTION
**Pull request by gabrieljenik**

### CODE REVIEW

**Comment by Shnoulle** [comment]
We can not do a find , and if null return json empty array (or [unknow])

**Comment by Shnoulle** [comment]
We can not do a find , and if null return json empty array (or [unknow])

**Comment by gabrieljenik** [comment]
Not sure I follow. An update was missing for some users.
So querying for the missing column.
If actually missing, we add it.

**Comment by Shnoulle** [comment]
Oh yes, better look : lack of a column.

We do the `foreach ($archivedTableSettings as $archivedTableSetting) {` in original push too ?

**Comment by gabrieljenik** [comment]
Yes, that is kept the same as it was on the original db update.

**Review by Shnoulle — APPROVED** [review]

**Review by Shnoulle — APPROVED** [review]

**Review by Shnoulle — APPROVED** [review]

**Review by Shnoulle — APPROVED** [review]

**Comment by Shnoulle** [comment]
@gabrieljenik since i don't see any potential issue, please fix the conflict and merge :)

**Review by Shnoulle — APPROVED** [review]

**Comment by gabrieljenik** [comment]
@c_schmitz, this seem a major impact stuff. Should we fix (again) and merge?

File: application/helpers/update/updates/Update_489.php [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+class Update_489 extends DatabaseUpdateBase
+{
+    public function up()
+    {
+        $columnSchema = $this->db->getSchema()->getTable('{{archived_table_settings}}')->getColumn('attributes');
+        if ($columnSchema === null) {
+            $this->db->createCommand()->addColumn('{{archived_table_settings}}', 'attributes', 'text NULL');
+            $archivedTableSettings = \Yii::app()->db->createCommand("SELECT * FROM {{archived_table_settings}}")->queryAll();
+            foreach ($archivedTableSettings as $archivedTableSetting) {
+                if ($archivedTableSetting['tbl_type'] === 'token') {
```
Comment (c-schmitz) [inline application/helpers/update/updates/Update_489.php:17]: Why not put this condition in the query?

File: application/helpers/update/updates/Update_489.php [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+                    $this->db->createCommand()->update('{{archived_table_settings}}', ['attributes' => json_encode(['unknown'])], 'id = :id', ['id' => $archivedTableSetting['id']]);
```
Comment (c-schmitz) [inline application/helpers/update/updates/Update_489.php:18]: Why is the field 'attributes' filled with the value 'unknown' ? Is this value actually used somewhere?

Comment (c-schmitz) [reply application/helpers/update/updates/Update_489.php:18]: To be able to fix this for 5.x and 6.x this needs to be put into different DBVersions.
Could you actually see that this was a cause by a bug we released in the past?

**Review by c-schmitz — CHANGES_REQUESTED** [review]

File: application/helpers/update/updates/Update_489.php (hunk 2 above)
Comment (gabrieljenik) [reply application/helpers/update/updates/Update_489.php:18]: Let me recall about this and be back :) Thanks!

File: application/helpers/update/updates/Update_489.php (hunk 1 above)
Comment (gabrieljenik) [reply application/helpers/update/updates/Update_489.php:17]: That's how it is laid out on the update 450 and 453

File: application/helpers/update/updates/Update_489.php (hunk 2 above)
Comment (gabrieljenik) [reply application/helpers/update/updates/Update_489.php:18]: > Why is the field 'attributes' filled with the value 'unknown' ? Is this value actually used somewhere?

Honestly, don;t recall now. Should look further.
That's how it is laid out on the update 450 and 453. As we are re-applying a previous recipe, wanted to keep it consistent.

**Comment by tiborpacalat** [comment]
Please make 2 patches out of it, one for 5.x and one for 6.x.

**Comment by gabrieljenik** [comment]
Done for master. Let's merge this first and then create a new one for v5

**Pull request merged**